### PR TITLE
FINERACT-1722: Fix loan repayment schedule with multitranche

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

The is an issue in the Loan repayment schedule, when the Loan product is multidisbursed and the disbursed amount is different (lower or higher) than approved amount

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by galovics** [comment]
@josehernandezfintecheandomx tests are failing. Please check.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 1]
```diff
@@ -5240,6 +5231,25 @@ public ChangedTransactionDetail updateDisbursementDateAndAmountForTranche(final
         return changedTransactionDetail;
     }
 
+    public BigDecimal getPrincipalAmountForRepaymentSchedule() {
+        BigDecimal principalAmount = BigDecimal.ZERO;
+
+        if (isMultiDisburmentLoan() && isDisbursed()) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:5142]: @josehernandezfintecheandomx I think this logic is faulty if the loan is not a multi disbursement loan... 

It should rather use the totalDisbursedAmount and no check whether it is multi disbursement or not... 

What do you think?

Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:5142]: Also, i am a little bit hesitant here...if expected disbursement details were provided, we should rather show repayment schedule based on those total principal, no? @bharathcgowda

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java [hunk 2]
```diff
@@ -131,23 +133,31 @@ private AdjustedDateDetailsDTO getAdjustedDateDetailsDTO(final LocalDate dueRepa
     private AdjustedDateDetailsDTO recursivelyCheckNonWorkingDaysAndHolidaysAndWorkingDaysExemptionToGenerateNextRepaymentPeriodDate(
             final AdjustedDateDetailsDTO adjustedDateDetailsDTO, final LoanApplicationTerms loanApplicationTerms,
             final HolidayDetailDTO holidayDetailDTO, final boolean isFirstRepayment) {
-
-        checkAndUpdateWorkingDayIfRepaymentDateIsNonWorkingDay(adjustedDateDetailsDTO, holidayDetailDTO, loanApplicationTerms,
-                isFirstRepayment);
-
+        final Recur recur = CalendarUtils.getICalRecur(holidayDetailDTO.getWorkingDays().getRecurrence());
+        final boolean isSevenDaysWeek = (recur.getDayList().size() == 7); // 7 Seven days in the week
+        // If Workings days are not seven day week
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java:138]: @josehernandezfintecheandomx Can you please tell about this newly introduced condition?

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanRepaymentRescheduleAtDisbursementTest.java [hunk 3]
```diff
@@ -134,11 +134,11 @@ public void testLoanRepaymentRescheduleAtDisbursement() {
         Calendar date = Calendar.getInstance(Utils.getTimeZoneOfTenant());
         date.set(2015, Calendar.MARCH, 16);
         expectedvalues.put("dueDate", getDateAsArray(date, 0));
-        expectedvalues.put("principalDue", "834.71");
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanRepaymentRescheduleAtDisbursementTest.java]: @josehernandezfintecheandomx Could you please tell more why this test was changed?

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanReschedulingWithinCenterTest.java [hunk 4]
```diff
@@ -320,7 +320,7 @@ public void testCenterReschedulingMultiTrancheLoansWithInterestRecalculationEnab
 
         // VERIFY THE INTEREST
         Float interestDue = (Float) ((HashMap) loanRepaymnetSchedule.get(2)).get("interestDue");
-        assertEquals("41.05", String.valueOf(interestDue));
+        assertEquals("45.41", String.valueOf(interestDue));
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanReschedulingWithinCenterTest.java]: @josehernandezfintecheandomx Are you sure it is valid to change the interest in this test case?

**Review by adamsaghy — CHANGES_REQUESTED** [review]
@josehernandezfintecheandomx I am not 100% sure about we have the valid behaviour with this changes. Can we add a couple test case and involve @bharathcgowda to validate?

**Review by adamsaghy — COMMENTED** [review]
See my above concerns

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments inline

**Review by galovics — CHANGES_REQUESTED** [review]
Please cover your changes with tests @josehernandezfintecheandomx.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:5142]: Done for the Integration Test

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java:138]: It is for the Full week evaluation and avoid to process the next Working date If you have a full week

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanReschedulingWithinCenterTest.java (hunk 4 above)
Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanReschedulingWithinCenterTest.java]: It was set as original amount in the case

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanRepaymentRescheduleAtDisbursementTest.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanRepaymentRescheduleAtDisbursementTest.java]: It was set as original amount in the case

**Comment by josehernandezfintecheandomx** [comment]
> See my above concerns

The next test case was added as the original issue, disburse a lower amount than the approved amount

Screen Shot 2022-09-16 at 10 50 48

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/LoanApplicationTerms.java [hunk 5]
```diff
@@ -1702,6 +1588,18 @@ public Money getTotalDisbursedAmount() {
         return disbursedAmount;
     }
 
+    public Money getTotalMultiDisbursedAmount() {
+        Money disbursedAmount = Money.zero(getCurrency());
+        if (isMultiDisburseLoan()) {
+            for (DisbursementData disbursement : getDisbursementDatas()) {
+                disbursedAmount = disbursedAmount.plus(disbursement.getPrincipal());
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/LoanApplicationTerms.java:1595]: @josehernandezfintecheandomx Do we need to consider here the "not yet" disbursement details as well?
Many other places the disbursement details are filtered by whether they got already disbursed or it is just expected yet (actualDisbursementDate = null). What do you think?

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java [hunk 6]
```diff
@@ -483,4 +473,24 @@ private HashMap<String, String> collaterals(Integer collateralId, BigDecimal qua
...
+            log.info("Loan with {} periods", getLoanRepaymentSchedule.getPeriods().size());
+            for (GetLoansLoanIdRepaymentPeriod period : getLoanRepaymentSchedule.getPeriods()) {
+                if (period.getPeriod() != null) {
+                    log.info("Period number {} for due date {} and outstanding {} {}", period.getPeriod(), period.getDueDate(),
+                            period.getTotalOutstandingForPeriod(), period.getTotalInstallmentAmountForPeriod());
+                    if (period.getPeriod() == 1) {
+                        totalOutstandingForPeriod = period.getTotalOutstandingForPeriod();
+                        totalInstallmentAmountForPeriod = period.getTotalInstallmentAmountForPeriod();
+                    } else {
+                        assertTrue(Math.abs(period.getTotalOutstandingForPeriod() - totalOutstandingForPeriod) <= limit);
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java:489]: @josehernandezfintecheandomx Can you help me to understand why do we need "limit" to be introduced? In my understanding if we are testing whether the instalment amounts are equal or not... if there is any fractions still it should remains +-1, no?

**Review by galovics — APPROVED** [review]

**Pull request merged**